curp/src/sync_manager.rs

```diff
@@ -144,137 +139,233 @@ impl<C: Command + 'static> SyncManager<C> {
+            prev_log_index,
+            prev_log_term,
+            entries,
+            leader_commit,
+        ) {
+            Err(e) => {
+                error!("unable to serialize append entries request: {}", e);
+                return;
+            }
+            Ok(req) => req,
+        };
+
+        // send append_entries request and receive response
+        debug!("append_entries sent to {}", connect.addr);
+        let resp = connect
+            .append_entries(req, Self::APPEND_ENTRIES_TIMEOUT)
+            .await;
+
+        #[allow(clippy::unwrap_used)]
+        // indexing of `next_index` or `match_index` won't panic because we created an entry when initializing the server state
+        match resp {
+            Err(e) => warn!("Heartbeat error: {}", e),
+            Ok(resp) => {
+                let resp = resp.into_inner();
+                let mut state = state.write();
+                if resp.success {
+                    // update match_index and next_index
+                    *state.match_index.get_mut(&connect.addr).unwrap() = last_sent_index;
+                    *state.next_index.get_mut(&connect.addr).unwrap() = last_sent_index + 1;
+
+                    let majority_cnt = (state.others.len() + 1) / 2 + 1;
+                    // update commit_index if needed
+                    if state.commit_index < last_sent_index
+                        // If the majority of servers has replicated the log, commit
+                        // + 1 means including the leader itself
+                        && 1 + state
+                            .others
+                            .iter()
+                            .filter(|addr| state.match_index[*addr] >= last_sent_index)
+                            .count()
+                            >= majority_cnt
+                    {
+                        debug!("commit_index updated to {last_sent_index}");
+                        state.commit_index = last_sent_index;
+                        if let Err(e) = commit_trigger.send(()) {
+                            error!("ae_trigger closed: {}", e);
+                        }
+                        if let Err(e) = ae_trigger.send(()) {
+                            error!("ae_trigger closed: {}", e);
+                        }
+                    }
+                } else {
+                    // TODO: update term
+                    *state.next_index.get_mut(&connect.addr).unwrap() -= 1;
+                }
+            }
+        };
+    }
+
+    /// Background apply
+    async fn bg_apply<CE: 'static + CommandExecutor<C>>(
+        state: Arc<RwLock<State<C>>>,
+        cmd_executor: Arc<CE>,
+        comp_chan: SpmcKeyBasedSender<C::K, SyncCompleteMessage<C>>,
+        mut commit_notify: UnboundedReceiver<()>,
+        spec: Arc<Mutex<RawMutex, VecDeque<C>>>,
+    ) {
+        while commit_notify.recv().await.is_some() {
+            if state.read().need_commit() {
+                let mut state = state.write();
```

Comment:
Don't have to acquire this lock twice. Try [this](https://docs.rs/lock_api/latest/lock_api/struct.RwLock.html#method.upgradable_read).